txpipe/extensions/twopoint_scia.py

```diff
@@ -0,0 +1,502 @@
+from ..base_stage import PipelineStage
+from ..twopoint import TXTwoPoint 
+from ..data_types import HDFFile, ShearCatalog, TomographyCatalog, RandomsCatalog, FiducialCosmology, SACCFile, PhotozPDFFile, PNGFile, TextFile
+from ..utils.calibration_tools import apply_metacal_response, apply_lensfit_calibration 
+from ..utils.calibration_tools import read_shear_catalog_type
+import numpy as np
+import random
+import collections
+import sys
+
+# This creates a little mini-type, like a struct,
+# for holding individual measurements
+Measurement = collections.namedtuple(
+    'Measurement',
+    ['corr_type', 'object', 'i', 'j'])
+
+SHEAR_SHEAR = 0
+SHEAR_POS = 1
+POS_POS = 2
+SHEAR_POS_SELECT = 3
+
+class TXSelfCalibrationIA(TXTwoPoint):
+    """
+    This is the subclass of the Twopoint class that will handle calculating the
+    correlations needed for doing the Self-calibration Intrinsic alignment
+    estimation. 
+
+    It requires estimating 3d two-point correlations. We calculate the 
+    galaxy - galaxy lensing auto-correlation in each source bin.
+    We do this twice, once we add a selection-function, such that we only selects 
+    the pairs where we have the shear object in front of the object used for density. 
+    Once without imposing this selection funciton. 
+    """
+    name = 'TXSCIA'
+    inputs = [
+        ('shear_catalog', ShearCatalog),
+        ('shear_tomography_catalog', TomographyCatalog),
+        ('shear_photoz_stack', HDFFile),
+        ('lens_tomography_catalog', TomographyCatalog),
+        ('lens_photoz_stack', HDFFile),
+        ('random_cats_source', RandomsCatalog),
+        ('patch_centers', TextFile),
+        ('photoz_pdfs', PhotozPDFFile),
+        ('fiducial_cosmology', FiducialCosmology),
+        ('tracer_metadata', HDFFile),
+    ]
+    outputs = [
+        ('twopoint_data_SCIA', SACCFile),
+        ('gammaX_scia', SACCFile),
+    ]
+    # Add values to the config file that are not previously defined
+    config_options = {
+        'calcs':[0,1,2],
+        'min_sep':2.5,
+        'max_sep':250.,
+        'nbins':20,
+        'bin_slop':0.1,
+        'flip_g2':True,
+        'cores_per_task':20,
+        'verbose':1,
+        'source_bins':[-1],
+        'lens_bins':[-1],
+        'reduce_randoms_size':1.0,
+        'do_shear_pos': True,
+        'do_pos_pos': False,
+        'do_shear_shear': False, 
+        'var_method': 'jackknife',
+        '3Dcoords': True,
+        'metric': 'Rperp',
+        'use_true_shear': False,
+        'subtract_mean_shear':False,
+        'redshift_shearcatalog': False,
+        }
+
+    def run(self):
```

Comment:
I won't try to review the body of this code, as it's an extension project - I'll stick to the infrastructural stuff as I hope we'll get more extensions and at some point I wouldn't be able to keep track!

Reply:
I assigned two reviewers who know more about the extension project, so they can probably review the body of the code

Reply:
I'm just curious about the 'substract_mean_shear' flag does. When I managed to reproduce your plots, I didn't remove the mean shear from the bins at any point but we probably should discuss that at some point. In this case it didn't make a difference since the means essentially came out to be zero but it might make a difference in other cases.

Reply:
It is a flag for TwoPoint.py, so it is defined in the main class. line 421.